src/firejail/fs_trace.c

```diff
@@ -41,6 +41,27 @@ void fs_trace_preload(void) {
 		fclose(fp);
 		fs_logger("touch /etc/ld.so.preload");
 	}
+	if (arg_tracefile) {
+		if (arg_debug)
+			printf("Creating an empty trace log file: %s\n", arg_tracefile);
+		// create a bind mounted trace logfile that the sandbox can see
+		EUID_USER();
+		FILE *fp = fopen(arg_tracefile, "w");
+		if (!fp)
+			errExit("fopen");
+		SET_PERMS_STREAM(fp, firejail_uid, firejail_gid, S_IRUSR | S_IWRITE | S_IRGRP | S_IROTH);
+		fclose(fp);
+		EUID_ROOT();
+		fp = fopen(RUN_TRACE_FILE, "w");
+		if (!fp)
+			errExit("fopen " RUN_TRACE_FILE);
+		fclose(fp);
+		fs_logger2("touch ", arg_tracefile);
+		if (mount(arg_tracefile, RUN_TRACE_FILE, NULL, MS_BIND|MS_REC, NULL) < 0)
```

Comment:
There is a tocttou between the fopen and the actual mount. An evil user might succeed in replacing arg_tracefile just in time with a symbolic link (it would allow him to access files in all directories, no matter if he has search permission on the path or not, because we mount as root and mount follows symbolic links). 

In order to prevent this kind of symlink attacks, one trick is to open the resolved path as user with safe_fd(), run necessary or good-to-have checks on the file descriptor, and then mount using a symbolic link in /proc/self/fd. We are doing something like that already in a number of places, for example check out the x11_xorg() function.